src/ucp/rma/amo_sw.c

```diff
@@ -52,24 +52,31 @@ static size_t ucp_amo_sw_fetch_pack_cb(void *dest, void *arg)
     return ucp_amo_sw_pack(dest, arg, 1);
 }
 
-static ucs_status_t ucp_amo_sw_progress(uct_pending_req_t *self,
-                                        uct_pack_callback_t pack_cb, int fetch)
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_amo_sw_progress(uct_pending_req_t *self, uct_pack_callback_t pack_cb,
+                    int fetch)
 {
     ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
     ucp_ep_t *ep       = req->send.ep;
     ucs_status_t status;
     ssize_t packed_len;
 
+    /* make an asuumption here that EP was able to send the REQ, since there
+     * are transports (e.g. SELF - it does send-recv in the AM function) that is
+     * able to complete the remote request operation inside uct_ep_am_bcopy() */
+    ucp_ep_rma_remote_request_sent(ep);
     req->send.lane = ucp_ep_get_am_lane(ep);
-    packed_len = uct_ep_am_bcopy(ep->uct_eps[req->send.lane],
-                                 UCP_AM_ID_ATOMIC_REQ, pack_cb, req, 0);
+    packed_len     = uct_ep_am_bcopy(ep->uct_eps[req->send.lane],
+                                     UCP_AM_ID_ATOMIC_REQ, pack_cb, req, 0);
     if (packed_len > 0) {
-        ucp_ep_rma_remote_request_sent(ep);
         if (!fetch) {
             ucp_request_complete_send(req, UCS_OK);
         }
         return UCS_OK;
     } else {
+        /* unroll a marking remote request as sent, since uct_ep_am_bcopy()
+         * completed with error */
+        ucp_ep_rma_remote_request_unroll_sent(ep);
```

Comment:
need this also for rma_sw?

Reply:
SELF has UCT/RMA support, but better to add it

Reply:
also fixed the case when flush_state is invalid, so need to just do flush_state_ops_inc before `uct_ep_am_bcopy()`